File: Platforms/Common/Reign.Video.XNA/Buffers/BufferLayoutDesc.cs
using Microsoft.Xna.Framework.Graphics;
using Reign.Core;
using System;
using System.Collections.Generic;

namespace Reign.Video.XNA
{
	public class BufferLayoutDesc : BufferLayoutDescI
	{
		#region Properties
		internal VertexElement[] desc;
		#endregion

		#region Constructors
		public static BufferLayoutDesc New(List<BufferLayoutElement> elements)
		{
			return new BufferLayoutDesc(elements);
		}

		public static BufferLayoutDesc New(BufferLayoutTypes type)
		{
			return new BufferLayoutDesc(type);
		}

		public BufferLayoutDesc(List<BufferLayoutElement> elements)
		: base(elements)
		{
			init();
		}

		public BufferLayoutDesc(BufferLayoutTypes bufferLayoutType)
		: base(bufferLayoutType)
		{
			init();
		}

		private void init()
		{
			desc = new VertexElement[ElementCount];
			int i = 0;
			foreach (BufferLayoutElement element in Elements)
			{
				desc[i] = new VertexElement();
				switch (element.Usage)
				{
					case BufferLayoutElementUsages.Position: desc[i].VertexElementUsage = VertexElementUsage.Position; break;
					case BufferLayoutElementUsages.Color: desc[i].VertexElementUsage = VertexElementUsage.Color; break;
					case BufferLayoutElementUsages.UV: desc[i].VertexElementUsage = VertexElementUsage.TextureCoordinate; break;
					case BufferLayoutElementUsages.Normal: desc[i].VertexElementUsage = VertexElementUsage.Normal; break;
					case BufferLayoutElementUsages.Index: desc[i].VertexElementUsage = VertexElementUsage.BlendIndices; break;
					case BufferLayoutElementUsages.IndexClassic: desc[i].VertexElementUsage = VertexElementUsage.BlendIndices; break;
				}
				switch (element.Type)
				{
					case BufferLayoutElementTypes.Float: desc[i].VertexElementFormat = VertexElementFormat.Single; break;
					case BufferLayoutElementTypes.Vector2: desc[i].VertexElementFormat = VertexElementFormat.Vector2; break;
					case BufferLayoutElementTypes.Vector3: desc[i].VertexElementFormat = VertexElementFormat.Vector3; break;
					case BufferLayoutElementTypes.Vector4: desc[i].VertexElementFormat = VertexElementFormat.Vector4; break;
					case BufferLayoutElementTypes.RGBAx8: desc[i].VertexElementFormat = VertexElementFormat.Color; break;
				}
				desc[i].UsageIndex = (int)element.UsageIndex;
				desc[i].Offset = element.ByteOffset;
				++i;
			}
		}
		#endregion
	}
}

File: Platforms/Windows/Reign.Video.D3D9/BufferLayoutDesc.cpp
#include "pch.h"
#include "BufferLayout.h"

namespace Reign
{namespace Video
{namespace D3D9
{
	#pragma region Properties
	D3DVERTEXELEMENT9* BufferLayoutDesc::Desc::get() {return desc;}
	#pragma endregion

	#pragma region Constructors
	BufferLayoutDesc::BufferLayoutDesc(List<BufferLayoutElement>^ elements)
	: BufferLayoutDescI(elements)
	{
		init();
	}

	BufferLayoutDesc::BufferLayoutDesc(BufferLayoutTypes bufferLayoutType)
	: BufferLayoutDescI(bufferLayoutType)
	{
		init();
	}

	void BufferLayoutDesc::init()
	{
		null();
		try
		{
			desc = new D3DVERTEXELEMENT9[ElementCount+1];
			uint i = 0;
			for each(BufferLayoutElement element in Elements)
			{
				switch (element.Usage)
				{
					case (BufferLayoutElementUsages::Position): desc[i].Usage = D3DDECLUSAGE_POSITION; break;
					case (BufferLayoutElementUsages::Color): desc[i].Usage = D3DDECLUSAGE_COLOR; break;
					case (BufferLayoutElementUsages::UV): desc[i].Usage = D3DDECLUSAGE_TEXCOORD; break;
					case (BufferLayoutElementUsages::Normal): desc[i].Usage = D3DDECLUSAGE_NORMAL; break;
					case (BufferLayoutElementUsages::Index): desc[i].Usage = D3DDECLUSAGE_BLENDINDICES; break;
					case (BufferLayoutElementUsages::IndexClassic): desc[i].Usage = D3DDECLUSAGE_BLENDINDICES; break;
					default: Debug::ThrowError("BufferLayoutDesc", "Unsuported ElementUsage"); break;
				}

				switch (element.Type)
				{
					case (BufferLayoutElementTypes::Float): desc[i].Type = D3DDECLTYPE_FLOAT1; break;
					case (BufferLayoutElementTypes::Vector2): desc[i].Type = D3DDECLTYPE_FLOAT2; break;
					case (BufferLayoutElementTypes::Vector3): desc[i].Type = D3DDECLTYPE_FLOAT3; break;
					case (BufferLayoutElementTypes::Vector4): desc[i].Type = D3DDECLTYPE_FLOAT4; break;
					case (BufferLayoutElementTypes::RGBAx8): desc[i].Type = D3DDECLTYPE_D3DCOLOR; break;
					default: Debug::ThrowError("BufferLayoutDesc", "Unsuported ElementType"); break;
				}

				desc[i].UsageIndex = element.UsageIndex;
				desc[i].Offset = element.ByteOffset;
				desc[i].Method = D3DDECLMETHOD_DEFAULT;
				desc[i].Stream = element.StreamIndex;
				++i;
			}
			desc[ElementCount].Stream = 0xFF;
			desc[ElementCount].Offset = 0;
			desc[ElementCount].Type = D3DDECLTYPE_UNUSED;
			desc[ElementCount].Method = 0;
			desc[ElementCount].Usage = 0;
			desc[ElementCount].UsageIndex = 0;
		}
		catch (Exception^ ex)
		{
			delete this;
			throw ex;
		}
	}

	BufferLayoutDesc::~BufferLayoutDesc()
	{
		if (desc) delete desc;
		null();
	}

	void BufferLayoutDesc::null()
	{
		desc = 0;
	}
	#pragma endregion
}
}
}